apollo-router/src/plugins/rhai.rs

```diff
@@ -0,0 +1,654 @@
+use std::{path::PathBuf, str::FromStr, sync::Arc};
+
+use apollo_router_core::{
+    plugin_utils, Context, ExecutionRequest, ExecutionResponse, QueryPlannerRequest,
+    QueryPlannerResponse, SubgraphRequest, SubgraphResponse,
+};
+use apollo_router_core::{
+    register_plugin, Error, Object, Plugin, RouterRequest, RouterResponse, Value,
+};
+use futures::executor::block_on;
+use http::HeaderMap;
+use http::{header::HeaderName, HeaderValue};
+use rhai::serde::{from_dynamic, to_dynamic};
+use rhai::{Dynamic, Engine, EvalAltResult, Scope, AST};
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde_json_bytes::ByteString;
+
+use tower::{util::BoxService, BoxError, ServiceExt};
+
+const CONTEXT_ERROR: &str = "__rhai_error";
+
+macro_rules! service_handle_response {
+    ($self: expr, $service: expr, $function_name: expr, $response_ty: ident) => {
+        let this = $self.clone();
+        let function_found = $self
+            .ast
+            .iter_fn_def()
+            .any(|fn_def| fn_def.name == $function_name);
+        $service = $service
+            .map_response(move |mut response: $response_ty| {
+                let previous_err = block_on(async {
```

Comment:
here too, block_on will probably create issues with the tokio executor. I think [and_then](https://docs.rs/tower/latest/tower/trait.ServiceExt.html#method.and_then) should be used instead of `map_response`